### DISCUSSION TITLE
Concurrency for Solo request

### DISCUSSION THREAD

**Discussion by jacksontong** [saloonphp/saloon#310] [Q&A] [answered]
Is it possible to make concurrent requests with [Solo Request](https://docs.saloon.dev/digging-deeper/solo-requests)? 

https://docs.saloon.dev/digging-deeper/concurrency-and-pools

**Comment by juse-less** [comment, marked as answer]
I believe it should be. At least in theory.
However, for true concurrency, you'd need to use the same HTTP client, to my knowledge.
At least from a cURL perspective (Guzzle default).
But I could be wrong here.

Nonetheless, from a SoloRequest perspective, I _think_ it should be quite easy.
But you could try to instantiate your own Pool, providing a Connector for all SoloRequests to run with (to use the same HTTP client).
See the [Pool class](https://github.com/saloonphp/saloon/blob/dc0284d6dcbe21ee7f4e0fe746a75c7431e7fba5/src/Http/Pool.php#L58) for constructor arguments.

Maybe something like the following (but untested):
```php
$pool = new Pool(
    // Same connector that individual SoloRequests use.
    // So let's re-use a new one for all SoloReuqests, to utilise the same HTTP client.
    connector: new NullConnector,

    // Either iterable or a callable that returns an iterable of SoloRequests.
    requests: function (): iterable {},

    // Default is 5 concurrent Requests.
    concurrency: 5,

    // Handle Responses for the SoloRequests, or replace the callable with null, if none is needed.
    responseHandler: function (Response $response, int|string $key, PromiseInterface $poolAggregate) {},

    // Handle Exceptions for the SoloRequests, or replace the callable with null, if none is needed.
    exceptionHandler: function (mixed $reason, int|string $key, PromiseInterface $poolAggregate) {}, 
);

// Then the regular flow applies.
$pool->send()->wait();
```

**Reply by Sammyjo20** [reply]
This is a great answer @juse-less - this is exactly how I would have approached concurrency for Solo Requests
